proposals/p1839.md

```diff
@@ -0,0 +1,331 @@
+  var x: [i32; :, :];
+  allocate(x, /*shape=*/(3, 2));
+  deallocate(x);
+```
+Calling of deallocation for non-allocated arrays leads to runtime error.
+
+### Operators
+
+Arrays may be modified in scalar and vector ways.
+
+1. Scalar way:
+  ```carbon
+    var x: [i32; 3; 2] = ((0, 1, 2), (3, 4, 5));
+    var y: auto = -x;
+    // each to each elements are summarized
+    var z: auto = x + y; // z = ((0,0,0), (0,0,0))
+  ```
+  If shapes are inconsistent, runtime error is.
+
+2. Vector way:
+  ```carbon
+    var x: [i32; 3; 2] = ((0, 1, 2), (3, 4, 5));
+    // multiply each element by 2
+    x *= 2; // x = ((0, 2, 4), (6, 8, 10));
+  ```
+
+### Slices and indexes
+
+Array slices (or sections) has the following syntax:
+```carbon
+  var x: [i32; 3; 2] = ((0, 1, 2), (3, 4, 5));
+  var z: auto = x[:, 1]; // z = (3, 4, 5)
+```
+
+Slice syntax: `a:b:s`. Here, `a` is the begin of slice, `b` is the end of slice
+and `s` is the step in slice. `:s` can be omitted, then it means that step is 1.
+If `a` is omitted, slice starts from the beginning of array. The same situation
+is for `b`, but slice ends at the end of arrays. `:` is the same as `::`.
+
+If array is inside indexes, it uses array values as indexes:
+```carbon
+  var ind: [i32; :] = (0, 1, 1, 2, 2);
+  var x: [i32; :] = (3, 5, 7);
+  var y: auto = x[ind]; // y = (3, 5, 5, 7, 7)
+```
+
+Slices can be also in left-hand side:
+```carbon
+  var x: [i32; 3; 2] = ((0, 1, 2), (3, 4, 5));
+  x[:, 0] = x[:, 1]; // x = ((3, 4, 5), (3, 4, 5));
+```
+
+### Iterators (?)
+
+In multidimensional arrays, it may be useful to have _iterators_ (row-major
+order):
+```carbon
+  var A: [f64; 2; 3; 4];
+  var it: auto = A[:, ...];
+  for(i: auto in it) { ... }
+```
+In this example, `i` presents `A[0,:,:]` and `A[1,:,:]` sequentially.
+Also, iterator may use last dimension (column-major order):
+```carbon
+  var A: [f64; 4; 3; 2];
+  var it: auto = A[..., :];
+  for(i: auto in it) { ... }
+```
+In this example, `i` presents `A[:,:,0]` and `A[:,:,1]` sequentially.
+In both cases, `i` is two dimensional array.
+
+`...` masks all dimensions.
+
+### Array generators
+
+Slice syntax can be used for constructing arrays:
+```carbon
+  var x: [i32; 3; 2] = ((0:4:2), (5:3:-1));
+  var y: [i32; 3; 2] = ((0, 2, 4), (5, 4, 3));
+```
+Here, `x` and `y` have the same values.
+
+### Functions
+
+Usually, arrays uses as is. Below, sum of two arrays is:
+```carbon
+  fn sum[T:! Type](x: T, y: T) -> T {
+    return x + y;
+  }
+```
+
+Function returning 1D array:
+```carbon
+  fn arr1D[T:! Type](x: T, y: T) -> T[:] {
+    return (x, y);
+  }
+```
+or 2D array:
+```carbon
+  fn arr2D[T:! Type](x: T, y: T) -> T[:,:] {
+    return ((x, x), (y, y));
+  }
+```
+
+Dimensions may be specified explicitly:
+```carbon
+  fn arr1D[T:! Type](x: T, y: T) -> T[2] {
+    return (x, y);
+  }
+```
+
+Lowering dimensions:
+```carbon
+  fn unarr[T:! Type](x: T[:], y: T[:]) -> T {
+    return sum(x + y);
+  }
+```
+
+#### Elemental functions
```

Comment:
I don't really understand the background and rationale for this feature. For example, what problems does it solve? Can those problems be solved with libraries instead of language features? Is there precedent for this feature in other languages?

This might be a good piece to split out into a separate proposal.